src/log.rs

```diff
@@ -81,6 +84,376 @@ type CompatFilterStream<T, R> = Compat01As03<FilterStream<T, R>>;
+            filter = filter.address(self.address);
+        }
+        if self.topics != TopicFilter::default() {
+            filter = filter.topic_filter(self.topics.compat())
+        }
+
+        filter
+    }
+
+    /// Returns a stream that resolves into a page of logs matching the filter
+    /// builder's parameters.
+    pub fn past_logs(self) -> PastLogStream<T> {
+        PastLogStream::from_builder(self)
+    }
+}
+
+/// A stream that emits past logs one page at a time.
+///
+/// Note this is a failure tolerant stream in that when a failure is encountered
+/// and `Some(Err(_))` is yielded, polling the stream again will resume from the
+/// last error and will not return `None`.
+#[must_use = "streams do nothing unless you poll them"]
+#[pin_project]
+pub struct PastLogStream<T: Transport> {
+    web3: Web3<T>,
+    /// The final `to_block`. This is used for the last page of past logs to
+    /// ensure that querying until "latest" works as expected in case of long
+    /// queries where new blocks get added during the process.
+    to_block: BlockNumber,
+    /// The block page size being used for queries.
+    block_page_size: u64,
+    /// The web3 filter used for retrieving the logs.
+    filter: FilterBuilder,
+    #[pin]
+    state: PastLogState<T>,
+}
+
+/// The inner state for the past log stream.
+#[pin_project]
+enum PastLogState<T: Transport> {
+    /// The stream is initilizing and will decide based on its block range if
+    /// the latest block number should be queried or it it can already start
+    /// retrieving log pages.
+    Init {
+        from_block: BlockNumber,
+    },
+    /// The block range is being determined by querying the current
+    RetrievingBlockRange {
+        start_block: u64,
+        #[pin]
+        end_block: MaybeReady<CompatCallFuture<T, U64>>,
+    },
+    StartRetrievingPage {
+        page_block: u64,
+        end_block: u64,
+    },
+    RetrievingPage {
+        page_block: u64,
+        end_block: u64,
+        #[pin]
+        page: CompatCallFuture<T, Vec<Log>>,
+    },
+    Done,
+}
+
+impl<T: Transport> PastLogStream<T> {
+    /// Creates a new past log stream from a filter builder.
+    pub fn from_builder(builder: LogFilterBuilder<T>) -> Self {
+        let web3 = builder.web3.clone();
+        let from_block = builder.from_block.unwrap_or(BlockNumber::Latest);
+        let to_block = builder.to_block.unwrap_or(BlockNumber::Latest);
+        let block_page_size = builder.block_page_size.unwrap_or(DEFAULT_BLOCK_PAGE_SIZE);
+
+        PastLogStream {
+            web3,
+            to_block,
+            filter: builder.into_filter(),
+            block_page_size,
+            state: PastLogState::Init { from_block },
+        }
+    }
+}
+
+impl<T: Transport> Stream for PastLogStream<T> {
+    type Item = Result<Vec<Log>, ExecutionError>;
+
+    #[project]
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```

Comment:
I wonder, could we write this instead as `async fn next() -> Result<Vec<Log, ExecutionError>` that works like it does for iterators? That way the implementation can be a lot simpler. Then when we have this function we can turn it into a stream in a couple of lines if we still want the stream or there might even be a function the futures crate that already does this.

Reply:
Interesting observation. I like when things are easily accessible!

Reply:
Yeah, this PR was from 26 days ago before moving to async/await, I had just forgotten to request a review.

This is a cool idea - I will re-implement with something like [`stream::unfold`](https://docs.rs/futures/0.3.5/futures/stream/fn.unfold.html)!